### PR TITLE
Event Hooks Overview

### PR DESCRIPTION
**Pull request by motikagan-okta**
## Description:
- Conceptual overview of Event Hooks.
- Event Hooks functionality is  currently targeted for EA release in 2019.04.0.

### Resolves:

* [OKTA-209169](https://oktainc.atlassian.net/browse/OKTA-209169)

### CODE REVIEW

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+To handle event hook calls from Okta, you need to implement a web service with an Internet-accessible endpoint. It's your responsibility to develop the code and to arrange its hosting on a system external to Okta. Okta defines the REST API contract for the requests that it will send.
+
+Event hooks are related to, but different from, Okta [inline hooks](/use_cases/inline_hooks/). Event hooks are meant to provide information about events that occurred, not offer a way to affect the underlying Okta process flow. For that, you need to use inline hooks. Also, unlike inline hooks, event hooks are asynchronous calls, meaning that the process flow that triggered the event hook continues without stopping or waiting for any response from your external service.
+
+## Which Events are Eligible?
+
+During the initial configuration procedure for an event hook, you specify which event types you want the event hook to deliver. The event types that can be specified are a subset of the event types that the Okta System Log captures. To see the list of event types currently eligible for use in event hooks, query the Event Types catalog with the query parameter `webhook-eligible`:
+
```
Comment (amoghvasekar-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: This has been renamed to `event-hook-eligible` to be consistent with the naming for feature

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+### One-Time Verification Request
+
+After registering an event hook, but before you can use it, you need to have Okta make a one-time GET verification request to your endpoint, passing your service a verification value that it needs to send back. This serves as a test confirming that you control the endpoint.
+
+This one-time verification request is the only GET request Okta will send to your external service, while the ongoing requests to notify your service of event occurrences will be HTTPS POST requests. Your web service can use the GET versus POST distinction to implement logic to handle this special one-time request.
+
+The way your service needs to handle this one-time verification is as follows: The request from Okta will contain an HTTP header named `X-Okta-Verification-Token`. Your service needs to read the value of that header and return it in the response body, in a JSON object named `verification`, i.e.: `{ "verification" : "value_from_header" }`. Note thate the value comes to you in an HTTP header, but you need to send it back in a JSON object.
```
Comment (amoghvasekar-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: the header has been renamed to `X-Okta-Verification-Challenge` after feedback.
Also, nit : typo `thate`

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+### Ongoing Event Delivery
+
+After verification, for ongoing notification of events, Okta sends HTTPS POST requests to your service. The JSON payload of these requests is where Okta provides specific information about events that have occurred.
+
+Information is encapsulated in the JSON payload in the `data.events` object. The `data.events` object is an array, in order to allow sending multiple events in a single POST request. Events that occur within a short time of each other will be amalgamated, with each each element of the array providing information on one event.
+
+ The content of each array element is an object of the [LogEvent](/docs/api/resources/system_log/#example-logevent-object) type. This is the same object that the [System Log API](/docs/api/resources/system_log/) defines for System Log events. See the documentation there for information on the object and its sub-objects.
+
+ Delivery of events is best effort. Delivery may be delayed by network conditions, and, in some cases, multiple requests can arrive at the same time after a delay.
```
Comment (amoghvasekar-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: I'd suggest adding a couple of things - 
1. Delivery of events is best effort. `Events are delivered at-least once`.
2. `Events may be delivered out-of-order in some cases. Okta recommends using the `published` attribute of the event object for strict ordering, if needed.`

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+Information is encapsulated in the JSON payload in the `data.events` object. The `data.events` object is an array, in order to allow sending multiple events in a single POST request. Events that occur within a short time of each other will be amalgamated, with each each element of the array providing information on one event.
+
+ The content of each array element is an object of the [LogEvent](/docs/api/resources/system_log/#example-logevent-object) type. This is the same object that the [System Log API](/docs/api/resources/system_log/) defines for System Log events. See the documentation there for information on the object and its sub-objects.
+
+ Delivery of events is best effort. Delivery may be delayed by network conditions, and, in some cases, multiple requests can arrive at the same time after a delay. 
+
+### Timeout and Retry
+
+When Okta calls your external service, it enforces a default timeout of 3 seconds. Okta will attempt at most one retry.
```
Comment (amoghvasekar-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Similar to inline hooks - `Response with a 4xx status code is not retried`

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 1 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Updated.

Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Updated.

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 2 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Updated.

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 4 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Added this sentence on 4xx responses not retried.

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 3 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Changed this paragraph to:
"Delivery of events is best effort. Events are delivered at least once. If you have more than one event hook configured to deliver the same event, it will only be sent by one of them. Delivery may be delayed by network conditions. In some cases, multiple requests may arrive at the same time after a delay, or events may arrive out of order. To establish ordering, you can use the time stamp contained in the `data.events.published` property of each event."

Comment (amoghvasekar-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: `If you have more than one event hook configured to deliver the same event, it will only be sent by one of them.` -> We're actually fixing this, and allowing sending same event to multiple hooks.

Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Have now removed the statement that said "If you have more than one event hook configured to deliver the same event, it will only be sent by one of them."

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
+---
+title: Event Hooks
+excerpt: Use Okta events to drive custom process flows.
+---
+
+# Event Hooks
```
Comment (amoghvasekar-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: A few more things we should add in general - 
1. Rate limits. Currently, up to a maximum of 100k events are delivered in a 24hr period.
2. Limit on number of event hooks that can be created (10).
3. Okta currently does not provide any SLA for time between event publication and event delivery. @ankurdatta-okta can clarify more here

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+- the URI of your external service's endpoint
+- the list of specific event types you want to use the event hook for
+- the secret value Okta should send in the authorization header of requests
+
+The response you receive will confirm creation of the event hook and provide you with the unique ID value of the created event hook.
+
+### Verifying an Event Hook
+
+After registering the event hook, you need to trigger the one-time verification process by making a POST request to the `/api/v1/eventHooks/${eventHookId}/verify` API. Okta will then make the GET request to your endpoint. If verification is successful, the JSON payload of the response to your call to the `/verify` API will contain a property called `verfication` set to `VERIFIED`.
```
Comment (amoghvasekar-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: the property has been renamed to `verificationStatus`

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 5 above)
Comment (amoghvasekar-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Adding things so I don't forget - 
4. Best Practices? (especially around not waiting for workflow to complete before returning response to Okta)

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+excerpt: Use Okta events to drive custom process flows.
+---
+
+# Event Hooks
+
+<ApiLifecycle access="ea" />
+
+## What Are Okta Event Hooks?
+
+Event hooks are outbound calls from Okta, sent when specified events occur in your org. They take the form of HTTPS REST calls to a URL you specify, encapsulating information about the events in JSON objects in the request body. These calls from Okta are meant to be used as triggers for process flows within your own software systems.
```
Comment (ankurdatta-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: It would be nice if we can mention somewhere that Event Hooks are Okta’s **Webhooks** offering. We need to have one mention of the word "Webhooks" because this will improve the discoverability of the Event Hooks developer docs pages.

Comment (ankurdatta-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Let us also include one example use case on when Event Hooks can be used. We can use this example snippet and reword it:

> For example, one can choose to subscribe to user deactivation events in Okta. When this event occur, Okta will notify your endpoint with additional information about that deactivation event. You can write custom code to have this event trigger a downstream workflow such as send an email alert, create a ticket, modify a record in their CRM system, etc.

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+To handle event hook calls from Okta, you need to implement a web service with an Internet-accessible endpoint. It's your responsibility to develop the code and to arrange its hosting on a system external to Okta. Okta defines the REST API contract for the requests that it will send.
+
+Event hooks are related to, but different from, Okta [inline hooks](/use_cases/inline_hooks/). Event hooks are meant to provide information about events that occurred, not offer a way to affect the underlying Okta process flow. For that, you need to use inline hooks. Also, unlike inline hooks, event hooks are asynchronous calls, meaning that the process flow that triggered the event hook continues without stopping or waiting for any response from your external service.
+
+## Which Events are Eligible?
+
+During the initial configuration procedure for an event hook, you specify which event types you want the event hook to deliver. The event types that can be specified are a subset of the event types that the Okta System Log captures. To see the list of event types currently eligible for use in event hooks, query the Event Types catalog with the query parameter `event-hook-eligible`:
```
Comment (ankurdatta-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: After the sentence _"The event types that can be specified are a subset of the event types that the Okta System Log captures"_ and before the sentence _"To see the list of event types currently eligible...."_, it would be nice to briefly mention which categories of events are available. For example, we can mention something such as this:

> These event types can include events describing the lifecycle state changes or modifications of core Okta objects such as User, Application, Groups etc. or events describing important milestones during user flows in Okta such as user logged in, user password reset etc. To see the list of event types currently eligible....

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 7 above)
Comment (ankurdatta-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Based on beta feedback and recent customer conversations, we should briefly mention somewhere that in the past, customers had only one option to track events which was to poll for these events using the SystemLog API. So, now with Event Hooks, Okta will push the events only when they occur. This is significantly convenient for the events which are less infrequent yet very urgent.

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+Information is encapsulated in the JSON payload in the `data.events` object. The `data.events` object is an array, in order to allow sending multiple events in a single POST request. Events that occur within a short time of each other will be amalgamated, with each each element of the array providing information on one event.
+
+ The content of each array element is an object of the [LogEvent](/docs/api/resources/system_log/#example-logevent-object) type. This is the same object that the [System Log API](/docs/api/resources/system_log/) defines for System Log events. See the documentation there for information on the object and its sub-objects.
+
+ Delivery of events is best effort. Events are delivered at least once. Delivery may be delayed by network conditions. In some cases, multiple requests may arrive at the same time after a delay, or events may arrive out of order. To establish ordering, you can use the time stamp contained in the `data.events.published` property of each event.
+
```
Comment (ankurdatta-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: This section on "Requests sent by Okta" is very well written :).  

Under "Ongoing Event Delivery", we need to include details on how to troubleshoot in the event of event hook delivery failure (this can possibly be a separate section). For now, we can mention something on the lines of this:

> As mentioned, there is a very slight possibility that the event might not get delivered or may be delayed due to network conditions. We recommend checking the System Log event event_hook.delivery to identify delivery failures. To fetch missing events, we suggest using the System Log API to pull the undelivered events.

Comment (ankurdatta-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: I noticed that, in general, we do not list the Syslog events for a feature in the developer docs for that feature. But it would be nice if we start doing so. IMHO, without knowing which Syslog events to track, the customers will find it even more difficult to troubleshoot. So, at least for Event Hooks, we should list the related events or ask customers to visit the Event Types link 
https://developer.okta.com/docs/api/resources/event-types/?q=event_hook

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+### Your Service's Responses to Event Delivery Requests
+
+Your external service's responses to Okta's ongoing event delivery POST requests should all be empty, and should have an HTTP status code of 200 (OK) or 204 (No Content).
+
+## Event Hook Setup
+
+### Registering an Event Hook
+
+After implementing your external service, you need to tell Okta it exists and configure it, by making a POST request to the `/api/v1/eventHooks` API. You use the JSON payload of that request to provide information on the event hook you are registering, including:
```
Comment (ankurdatta-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Very minor comment: The "it" modifier is dangling because there are two subjects - "external service" and "Okta". I would suggest rewording this sentence as follows:

> After implementing your external service, you need to tell Okta that this service exists. You can do this configuration by simply making a POST request to the `/api/v1/eventHooks` API.

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+### One-Time Verification Request
+
+After registering an event hook, but before you can use it, you need to have Okta make a one-time GET verification request to your endpoint, passing your service a verification value that it needs to send back. This serves as a test confirming that you control the endpoint.
+
+This one-time verification request is the only GET request Okta will send to your external service, while the ongoing requests to notify your service of event occurrences will be HTTPS POST requests. Your web service can use the GET versus POST distinction to implement logic to handle this special one-time request.
+
+The way your service needs to handle this one-time verification is as follows: The request from Okta will contain an HTTP header named `X-Okta-Verification-Challenge`. Your service needs to read the value of that header and return it in the response body, in a JSON object named `verification`, i.e.: `{ "verification" : "value_from_header" }`. Note that the value comes to you in an HTTP header, but you need to send it back in a JSON object.
```
Comment (ankurdatta-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Under the sub-section "One-Time Verification Request", the last two paragraphs contain implementation related info. IMHO, it might be better to move these two paragraphs in under "Verifying an Event Hook" (in the section _Event Hooks Set Up_) because that is where we talk about how to make a POST request to the /api/v1/eventHooks/${eventHookId}/verify API.

We can still keep the first paragraph in the sub-section "One-Time Verification Request" where we describe why we do this one-time verification. At the end of this paragraph, we can simply add a link to the upcoming section "Verifying an Event Hook" on how to implement verification.

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 6 above)
Comment (ankurdatta-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: During the beta program, some of the less technical admins who are new to the world of webhooks, were unsure how these hooks get triggered and send the events. So, it would be great if we add a small section called "Triggering an Event Hook". This section will include just two lines. E.g.

> When the specific events configured in your Event Hook occur in Okta, the Event Hook will get triggered. You will receive a delivery payload as shown in the section below "Sample Event Delivery Payload".

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+- the URI of your external service's endpoint
+- the list of specific event types you want to use the event hook for
+- the secret value Okta should send in the authorization header of requests
+
+The response you receive will confirm creation of the event hook and provide you with the unique ID value of the created event hook.
+
+### Verifying an Event Hook
+
+After registering the event hook, you need to trigger the one-time verification process by making a POST request to the `/api/v1/eventHooks/${eventHookId}/verify` API. Okta will then make the GET request to your endpoint. If verification is successful, the JSON payload of the response to your call to the `/verify` API will contain a property called `verfication` set to `VERIFIED`.
+
```
Comment (ankurdatta-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Recently, we introduced lifecycle APIs for an Event Hook (i.e. activate/deactivate) - https://oktainc.atlassian.net/browse/OKTA-202357

So, we need a small section called "Managing the lifecycle of an Event Hook". This section would appear after the section on "Triggering an Event Hook" and would contain the following sentences:

> 

- When a new Event Hook is created, it's status is ACTIVE by default. 
- One can deactivate an active Event Hook by making a POST request to the /api/v1/eventHooks/${eventHookId}/lifecycle/deactivate API. This would change the status of the event hook to INACTIVE. 
- In order to fire this inactive Event Hook, one needs to activate the hook again by making a POST request to the /api/v1/eventHooks/${eventHookId}/lifecycle/activate API. While this will change the Event Hook's status to ACTIVE again, however due to security reasons, Okta requires you to verify this newly reactivated event hook once again. Please follow the steps mentioned in "Verifying an Event Hook" on how to verify an event hook.

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 6 above)
Comment (gabewerman-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: ```/api/v1/eventHooks/${eventHookId}/verify``` has been changed to ```/api/v1/eventHooks/${eventHookId}/lifecycle/verify```

**Comment by amoghvasekar-okta** [comment]
@susanharper-okta - This feature will go EA in 2019.04.1. Does that affect when you merge the changes?

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 12 above)
Comment (ankurdatta-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: We also need to mention that at a time an org can have a max number of 10 ACTIVE event hook configurations.

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+During the initial configuration procedure for an event hook, you specify which event types you want the event hook to deliver. The event types that can be specified are a subset of the event types that the Okta System Log captures. To see the list of event types currently eligible for use in event hooks, query the Event Types catalog with the query parameter `event-hook-eligible`:
+
+[https://developer.okta.com/docs/api/resources/event-types/?q=event-hook-eligible](/docs/api/resources/event-types/?q=event-hook-eligible)
+
+## Requests Sent by Okta
+
+The requests sent from Okta to your external service are HTTPS requests. POST requests are used for the ongoing delivery of events, and a one-time GET request is used for verifying your endpoint.
+
+### One-Time Verification Request
```
Comment (amoghvasekar-okta) [inline packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: In addition - it may take upto a minute for the event hook to used by Okta after it has been verified.

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 11 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: @ankurdatta-okta, customers have to code logic into their service to handle this one-time verification request, because the same service that handles ongoing requests will have to be able to handle this one too.

The "Verifying an Event Hook" procedure, on the other hand, is manually performed by a human being.

Isn't it clearer to keep those two things separate: Logic you need to implement in your code as you develop your software, versus setup steps you need to perform manually?

Comment (ankurdatta-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Sounds good. I think I understood now why you kept the two things separate. It would be nice if we add a hyperlink to the section on "Verifying an Event Hook" towards the end of the section on "One-Time Verification Request". For someone unfamiliar with the verification flow, this will help them relate the two separate steps.

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 6 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Updated (all changes in new pull request: https://github.com/okta/okta-developer-docs/pull/227)

Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Updated (changes are in new pull request: https://github.com/okta/okta-developer-docs/pull/227)

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 11 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Added a link from "One-Time Verification Request" section to "Verifying an Event Hook" section (all changes are in new pull request: https://github.com/okta/okta-developer-docs/pull/227)

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 13 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Added a sentence: "After your event hook is verified, it can take up to a minute for delivery of events to your endpoint to begin." (All changes are in new pull request: https://github.com/okta/okta-developer-docs/pull/227)

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 5 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Added:
- New section "Rate Limits", saying "Event hooks are limited to sending 100 000 events per 24-hour period."
- Addition to section "What are Okta Event Hooks?", saying: "You can create a maximum of ten event hooks in your org. Each event hook can be configured to deliver multiple event types."
- Addition to "Ongoing Event Delivery" section, saying: "No guarantee of maximum delay between event occurrence and delivery is currently specified."
 - Addition to "Your Service's Responses to Event Delivery Requests" section, saying: "As a best practice, you should return the HTTP response immediately, rather than waiting for any of your own internal process flows triggered by the event to complete."

(All changes are in new pull request: https://github.com/okta/okta-developer-docs/pull/227)

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 12 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: I think it would make for explanation of lifecycle operations model for the management API in the reference page for the API.

Comment (ankurdatta-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Sounds good. This is covered in Amogh's PR on the management API -- https://github.com/okta/okta-developer-docs/pull/193

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 10 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: Changed to: "After implementing your external service, you need to register it with Okta by making a POST request to the `/api/v1/eventHooks` API." (All changes are in new pull request: https://github.com/okta/okta-developer-docs/pull/227)

File: packages/@okta/vuepress-site/use_cases/event_hooks/index.md (hunk 9 above)
Comment (motikagan-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: @ankurdatta-okta, I'm hesitating regarding adding your recommendation to check the System Log API. If this is a case of a developer debugging their event hook handler, wouldn't it be easier for them to go to the System Log UI in Admin Console? But, if you mean adding code in the their implementation, to regularly poll the System Log API, doesn't that defeat the purpose of Event Hooks?

Comment (ankurdatta-okta) [reply packages/@okta/vuepress-site/use_cases/event_hooks/index.md]: > @ankurdatta-okta, I'm hesitating regarding adding your recommendation to check the System Log API. If this is a case of a developer debugging their event hook handler, wouldn't it be easier for them to go to the System Log UI in Admin Console? But, if you mean the developer adding code in the implementation, to regularly poll the System Log API, doesn't that defeat the purpose of Event Hooks?

I understand and agree with your concern. I'll provide some more context which will hopefully clarify things. 

Some time later this year, the Extensibility team will start supporting a feature called "Replay" which would allow Okta to go back to the history of event hooks delivered in the past and then allow the customer to ask Okta to resend the hooks for which there were delivery failures. At present, we do not support this functionality. As a result, we have to ask customers to use the System Log API to poll for events when the events get missed or undelivered. 

--
An example of a Replay feature:
https://pinpayments.com/developers/api-reference/webhooks#put-webhook-replay
